### PR TITLE
Giant cleanup of monattacks

### PR DESCRIPTION
**Pull request by Coolthulhu**
Finally removed most of the ugly casts of `target` to `player` in monattack.cpp. I encapsulated the `player->practice( "dodge", z->type->melee_skill )` into a creature function that also triggers `ma_ondodge`. And a second function that triggers on a hit that lands.

As a result, dodging bite attacks, smashes etc. is more similar to dodging melee attacks. It doesn't eat up dodge counter (it shouldn't really - dodge counter shouldn't waste chances at avoiding a smash or impale).
Getting hit by those attacks will now trigger defenses. For example, zombies can no longer avoid riot bot's zapback by biting it (they could do that before).

Cleaned up monsters attacking other monsters. They used different function (probably very old) that ignored cut attack damage, used different rolls etc. It still needs some more cleanup and unifying with the player version, but is way more proper.

Fixed a bug that would result in debug message spam whenever a monster is smashed into a monster, kills it, but then lands on one of its "children" spawned after parent's death. Currently more like hacked around it, but it's still way better than having the debug spam.

Fixed a rare bug where player with judo would get zapped by non-zapback monsters coming from up/down stairs.

Cleaned up (and/or made 3D) a lot of monster abilities. Mostly just changing coords from 2D to 3D and `player` to `Creature` (where possible), but did a more proper cleanup of:
- Acidic zombie barf
- Corrosive zombie spit and defense
- Bite
- Ranged bite/slap (from hollow zed)
- Impale
- Smoker

Rewrote a part of copbot function. It now shocks stuff that isn't player.

Removed vortex function. It was bugged (crashed the game whenever invoked) and fixing it isn't worth the effort. It was invoked only by one very rare monster, it was heavily bloated and full of special cases and it was pure magic at work so not lore-friendly in the slightest. Every single thing about this function was a reason to cut it out.
Left vortices as monsters that just stumble around.

### CODE REVIEW

**Comment by kevingranade** [comment]
Does this fix the issue where turrets firing underground drop shell casings on the surface?  It was just brought to my attention and it seemed like something in monattack was a likely culprit.

**Comment by Coolthulhu** [comment]
I think it should (didn't test this particular issue, though).
The problem was that the fake NPC that turrets use to shoot didn't get its z-pos set.

File: data/json/monsters.json [hunk 1]
```diff
@@ -4658,7 +4658,6 @@
     "luminance":0,
     "hp":20,
     "death_function":"MELT",
-    "special_attacks":[["VORTEX", 6]],
```
Comment (BevapDin) [inline data/json/monsters.json]: Out of curiosity, why did you remove that attack?

Comment (Coolthulhu) [inline data/json/monsters.json]: I explained it in the PR description. It's a big, buggy (crashes the game), outdated 2D function only used for a single very rare monster. And now there's an excuse not to port it over.

Comment (BevapDin) [inline data/json/monsters.json]: Oh, yes, my fault, reading text (as opposed to code) is not my strong side.

**Comment by BevapDin** [comment]
The handling of uncanny dodge vs normal dodge is slightly inconsistent:
```C++
{
    bool uncanny = target->uncanny_dodge(); // Note: may already print a message
    if( uncanny || dodge_check( z, target ) ) {
        target->add_msg_player_or_npc( m_warning, _("The %s lunges at you, but you dodge!"),
                                                  _("The %s lunges at <npcname>, but they dodge!"),
                                       z->name().c_str() );
        if( !uncanny ) {
            target->on_dodge( z, z->type->melee_skill * 2 );
        }
```
This will give two messages (one from dodging via the bionic (`uncanny_dodge`) and one from the standard dodging here). Other functions show only one message, like:
```C++
{
    if( target->uncanny_dodge() ) {
        return;
    }

    if( dodge_check( z, target ) ) {
        target->add_msg_player_or_npc( _("The %s takes a powerful swing at you, but you dodge it!"),
                                       _("The %s takes a powerful swing at <npcname>, who dodges it!"),
                                       z->name().c_str() );
```

**Comment by kevingranade** [comment]
This does not seem to fix the issue, visit a lab and most likely you'll see casings spawn on the surface and get messages about monsters fighting underground.
You'll also hear gunfire, but that's fine :D

**Comment by Coolthulhu** [comment]
Must be ranged code then, I'll check that out
